Reading & Writing wav with 32-bit resolution

Sidelobe opened the pull request:
I added  support for reading and writing of wave files with 32bit resolution with corresponding tests.

The unit tests currently use "hard-coded" reference data (got it through Matlab's `audioread`) and compare the first 64 samples of audio.

Also, I'm creating the output directory for the unit test in case it doesn't exist yet. Added it to .gitignore as well.

I haven't had time to look into the headers you use as reference, but guess `makeHeaders.py` is what I'm looking for.

adamstark commented:
Thanks for this! At first glance it looks great - I'll need to find some time to sit down and look properly but I'll make sure this gets added to the library at some point soon! :)

adamstark commented on AudioFile.cpp:
```diff
@@ -592,7 +598,7 @@ bool AudioFile<T>::saveToWaveFile (std::string filePath)
             else if (bitDepth == 24)
             {
-                int32_t sampleAsIntAgain = (int32_t) (samples[channel][i] * (T)8388608.);
+                int32_t sampleAsIntAgain = (int32_t) (samples[channel][i] * (T)(1<<24));
```
Should this be `1 << 23` rather than `1 << 24` ? 

8388608 is 1 << 23 which would match the hardcoded value in the line it replaces? :)

Sidelobe replied:
Absolutely!  stupid typo on my side ...
should be `1 << 23`
I had focused too much on the wav reading part and forgot to check the writing..
Updated the code 👍

adamstark commented on AudioFile.cpp:
```diff
@@ -603,6 +609,21 @@ bool AudioFile<T>::saveToWaveFile (std::string filePath)
             }
+            else if (bitDepth == 32)
+            {
+              int32_t sampleAsIntAgain = (int32_t) (samples[channel][i] * (T)(1<<32));
```
Need to look in more detail when it isn't 1am but potentially an issue here with the bit shift - i.e. `1 << 32` is larger than a `int32_t` right? So this should be something that can be multiplied by 1 or -1 and still fit into the int32_t. So if a int32_t range is -2147483648 to 2147483647, maybe it should be multiply by 2147483647? This would be `INT_MAX` again as your above code uses, but probably best to hard code it here in case of some unforeseen future system that decides to have a different value for `INT_MAX`?

Sidelobe replied:
Absolutely valid point you make -- but this was the same kind of typo as above -- it's supposed to be `1 << 31` -- corrected now.

adamstark replied:
I think we need to be using the max value for an int32 here again. `1 << 31` is `-2147483648` which isn't the value we need here. 

i.e. a sample value of `1.0` would be represented as `-2147483648`, and `-1.0` would be `+2147483648` (which is out of bounds for `int32_t`).

I think it should be...

`int32_t sampleAsIntAgain = (int32_t) (samples[channel][i] * (T)std::numeric_limits<std::int32_t>::max());`

Sound right? :)

adamstark replied:
(This thread is the `1 << 31` thing ^^^)

Sidelobe replied:
Oh yes, you're absolutely right -- it should be `(1 << 31) - 1 = 2147483647`, which is `std::numeric_limits<std::int32_t>::max()`

I made the changes -- and double checked in the debugger that `+1.0` is mapped to `2147483647` and `-1.0` to `-2147483647`.

Thanks for your review!

Sidelobe commented on AudioFile.cpp:
```diff
@@ -351,12 +351,18 @@ bool AudioFile<T>::decodeWaveFile (std::vector<uint8_t>& fileData)
+              int32_t sampleAsInt = fourBytesToInt (fileData, sampleIndex);
+              T sample = (T)sampleAsInt / (T)(INT32_MAX);
+              samples[channel].push_back (sample);
+            }
```
As per your input in the other comment, I changed INT_MAX to INT32_MAXin the above. It makes a lot of sense to be 32-bit specific here.

adamstark replied:
I think we might be better here to use `std::numeric_limits<std::int32_t>::max()` as it is a bit more modern. Or to hardcode the value as it doesn't change. Depends how many people would try to use the library on older platforms (I'm pretty sure I've got some other C++11 stuff in here already though).

Sidelobe replied:
Personally, I hate "magic numbers", even if they don't change..  I fully agree on `std::numeric_limits<std::int32_t>::max()` - that is indeed the modern way and it conveys what the number is supposed to be 👍 

I think C++11 is a good baseline standard to aim for this project. I tried compiling for C++98 as a quick test, and got like 20 errors, mostly due to closing template brackets needing a space (`> >`).

adamstark replied:
Ok, excellent. I've marked one other issue above unresolved (the `1 << 31` thing) and then I can merge this in and  have a look :)

Sidelobe commented on tests/AudioFileTests/WavLoadingTests.cpp:
```diff
@@ -95,6 +95,95 @@ BOOST_AUTO_TEST_CASE (WavLoadingTests_Stereo_24bit_44100)
+    { 0.0498672723770142   ,  0.0476465225219727 },
+    { 0.0204880237579346   ,  0.0452275276184082 },
+  };
+
```
This style of "embedded" test data is maybe not the nicest way... it does bloat the test code a bit.
Also, I could change this to `LLLLLRRRRR` rather than the current `LRLRLRLRLR` if you prefer.

adamstark replied:
I haven't merged this in yet (mainly just to not having much time) - but when I do I'll take a look at the tests properly :)

Sidelobe commented:
BTW, I just noticed that 32-bit AIFFs are being generated in the tests as well, even though the code is not implemented yet --> I found the reason why and I'm opening a separate issue for it.

I've fixed this in this PR -- or do you want this fix in a separate PR?

I've also added a TODO for 32 bit AIFF reading/writing -- which I'm happy to take in the next PR ;-)

Sidelobe commented:
I guess this PR is now obsolete --

adamstark commented:
I've added 32-bit support on develop now but I just need to check it works as I expect... :)
